swirl_lm/boundary_condition/immersed_boundary_method.py

```diff
@@ -317,6 +328,507 @@ def _check_mesh_size(a, b):
+      #     node_dist_max_i * node_dist_i)
+      # )**0.5
+      # where gamma_i is the un-normalized interpolation weight of the i-th
+      # enclosing grid node for a given image point. The normalized weight is
+      # then obtained by
+      # w = (1 / denom) * ∑_k gamma_k.
+      # Note that we always exclude the body intercept weight for Neumann boundary
+      # conditions. This is because the value AT the body intercept is not known
+      # before executing the interpolation of the enclosing grid nodes onto the
+      # image point. Consequently, we need a separate set of interpolation weights
+      # for Neumann boundary conditions.
+      # Reference: R. Franke, Scattered data interpolation: tests of some
+      # methods, Mathematics of Computation 38(157), 1982,
+      # https://www.ams.org/mcom/1982-38-157/S0025-5718-1982-0637296-4/
+      # And:
+      # Cai et al., JCP 429, 2021, specifically Eq. 35, p. 9
+
+      # For Dirichlet boundary conditions:
+      gamma = tf.nest.map_structure(
+          lambda _w: tf.math.divide(
+              node_dist_max - _w, node_dist_max * _w + _EPSILON)**_EXP,
+          node_dist,
+      )
+      denom = tf.math.reduce_sum(gamma, axis=0)
+
+      # For Neumann boundary conditions:
+      gamma_n = tf.nest.map_structure(
+          lambda _w: tf.math.divide(
+              node_dist_max - _w, node_dist_max * _w + _EPSILON)**_EXP,
+          node_dist[0:8],
+      )
+      denom_n = tf.math.reduce_sum(gamma_n, axis=0)
+
+      output = {}
+
+      output.update(
+          {
+              'w':
+                  tf.nest.map_structure(lambda _g: tf.math.divide(_g, denom),
+                                        gamma),
+              'w_neumann':
+                  tf.nest.map_structure(lambda _g: tf.math.divide(_g, denom_n),
+                                        gamma_n),
+          }
+      )
+
+      return output
+
+
+    interp_helper = get_interp_weights([x_ip, y_ip, z_ip], [p, q, s])
+    w = interp_helper['w']
+    w_n = interp_helper['w_neumann']
+
+    interp_helper_2x = get_interp_weights([x_ip2, y_ip2, z_ip2], [p2, q2, s2])
+    w_n_2x = interp_helper_2x['w_neumann']
+
+    def convert_to_3d_tensor(vector, indices, dtype=tf.float32):
+      """Scatters a 1D `vector` into a 3D tensor at elements defined by
+      `indices`, and sets all other elements to zero."""
+      return common_ops.scatter(
+          vector,
+          indices,
+          shape=[params.nx, params.ny, params.nz],
+          dtype=dtype
+      )
+
+
+    output = {'gp_mask': gp_mask}
+    output.update({
+        f'w{n}': convert_to_3d_tensor(w[n], ijk_gp) for n in range(9)
+    })
+    output.update({
+        f'w_neumann{n}': convert_to_3d_tensor(w_n[n], ijk_gp,
+        ) for n in range(8)
+    })
+    output.update({
+        f'w_neumann_2x{n}': convert_to_3d_tensor(w_n_2x[n], ijk_gp,
+        ) for n in range(8)
+    })
+    output.update(
+        {
+            'nx': unit_normals_masked[..., 0],
```

Comment:
Let's try to eliminate the unused ones (or only keep then with the presence of a debugging flag).